src/software/geom/algorithms/end_in_obstacle_sample.cpp

```diff
@@ -0,0 +1,51 @@
+#include <random>
+#include "software/geom/algorithms/end_in_obstacle_sample.h"
+#include "software/geom/point.h"
+#include "software/ai/navigator/obstacle/obstacle.hpp"
+
+
+Point endInObstacleSample(const std::vector<ObstaclePtr> obstacles, const Point &destination, const Rectangle &navigable_area, double multiplier, double range)
+{
+    for (auto const &obstacle : obstacles) {
+        if (obstacle->contains(destination)) {
+            goto get_nearest;
+        }
+    }
+    return destination;
+
+get_nearest:
+    for (auto const &obstacle : obstacles) {
+        if (obstacle->contains(destination)) {
+
+            Point closest_point = obstacle->closestPoint(destination);
```

Comment:
I think we can just do `closestPoint` on the first obstacle it's contained within. I feel like chances are that if the `closestPoint` to the first obstacle is in another obstacle, the rest of the `closestPoint`s will also be in an obstacle.

In that case the logic here could probably be simplified. Since you're already doing a loop over all obstacles above and you know that one obstacle contains it if you've gotten this far. So you can just do `closestPoint` with that obstacle, and then check if the `closestPoint` is in another obstacle.

Logic of checking if a point is in an obstacle could be refactored to a separate function (could return an `std::optional<ObstaclePtr>`, since we're performing it multiple times.